sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/pubsub/PubsubUnboundedSink.java

```diff
@@ -201,8 +206,15 @@ public void processElement(
           break;
       }
 
+      // TODO(sjvanrossum): https://github.com/apache/beam/issues/31828
+      // NOTE: Null and empty ordering keys are treated as equivalent.
       @Nullable String topic = dynamicTopicFn.apply(element);
-      K key = keyFunction.apply(ThreadLocalRandom.current().nextInt(numShards), topic);
+      @Nullable String orderingKey = message.getOrderingKey();
+      int shard =
+          Strings.isNullOrEmpty(orderingKey)
+              ? ThreadLocalRandom.current().nextInt(numShards)
+              : Hashing.murmur3_32_fixed().hashString(orderingKey, StandardCharsets.UTF_8).asInt();
```

Comment:
in general using a huge # of keys hurts performance. It seems like it would be better to still limit to numShards but be deterministic shard. The user can still for now increase the numShards very high if needed for performance.

If we need more publishing paralellism that seems like it should be done below by just adding publishBatch to scheduled executor and then joining, not via key parallelism.

Reply:
Makes sense. I see we already group messages into different batches based on ordering key down in `WriterFn`. In other words, downstream steps don't rely on bundles being grouped by ordering key.

Do we necessarily need to shard by ordering key here? I wonder if we can revert this section and keep to `numShards` CC @sjvanrossum

Reply:
Agreed, that was the intention based on an earlier comment of yours, but I guess it slipped my mind to apply bucketing.

Reply:
@ahmedabu98 a deterministic shard number based on the ordering key ensures that the writers only operate on distinct key ranges, thus avoiding the creation multiple buffers per ordering key with potentially a small number of batched elements per bundle. Volume skews across keys is as much a problem for Beam as it is for Pub/Sub and does not fit well with the feature since throughput is capped at 1MB/s per ordering key.

One thing that does concern me is ending up with a skewed distribution of keys across buckets, but that could be fixed by tweaking the bucketing algorithm.